reco_utils/evaluation/spark_diversity_evaluation.py

```diff
@@ -0,0 +1,361 @@
+                .orderBy(self.item_col)
+            )
+        return self.df_item_novelty
+
+    def user_novelty(self):
+        """Calculate average item novelty for each user's recommendations.
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: user_col, user_novelty
+        """
+        if self.df_user_novelty is None:
+            self.df_item_novelty = self.item_novelty()
+            self.df_user_novelty = (
+                self.reco_df.join(self.df_item_novelty, on=self.item_col)
+                .groupBy(self.user_col)
+                .agg(F.mean("item_novelty").alias("user_novelty"))
+                .orderBy(self.user_col)
+            )
+        return self.df_user_novelty
+
+    def novelty(self):
+        """Calculate average novelty for recommendations across all users.
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: novelty
+        """
+        if self.df_novelty is None:
+            self.df_user_novelty = self.user_novelty()
+            self.df_novelty = self.df_user_novelty.agg(
+                F.mean("user_novelty").alias("novelty")
+            )
+        return self.df_novelty
+
+    # serendipity metrics
+    def user_item_serendipity(self):
+        """Calculate serendipity of each item in the recommendations for each user.
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: user_col, item_col, user_item_serendipity
+        """
+        # for every user_col, item_col in reco_df, join all interacted items from train_df.
+        # These interacted items are repeated for each item in reco_df for a specific user.
+        if self.df_user_item_serendipity is None:
+            self.df_cosine_similariy = self._get_cosine_similarity().orderBy("i1", "i2")
+            self.df_user_item_serendipity = (
+                self.reco_df.withColumn(
+                    "reco_item", F.col(self.item_col)
+                )  # duplicate item_col to keep
+                .select(
+                    self.user_col,
+                    "reco_item",
+                    F.col(self.item_col).alias("reco_item_tmp"),
+                )
+                .join(
+                    self.train_df.select(
+                        self.user_col, F.col(self.item_col).alias("train_item_tmp")
+                    ),
+                    on=[self.user_col],
+                )
+                .select(
+                    self.user_col,
+                    "reco_item",
+                    F.least(F.col("reco_item_tmp"), F.col("train_item_tmp")).alias(
+                        "i1"
+                    ),
+                    F.greatest(F.col("reco_item_tmp"), F.col("train_item_tmp")).alias(
+                        "i2"
+                    ),
+                )
+                .join(self.df_cosine_similariy, on=["i1", "i2"], how="left")
+                .fillna(0)
+                .groupBy(self.user_col, F.col("reco_item").alias(self.item_col))
```

Comment:
small simplification here
```suggestion
                self.reco_df.select(
                    self.user_col,
                    self.item_col,
                    F.col(self.item_col).alias("reco_item_tmp"),  # duplicate item_col to keep
                )
                .join(
                    self.train_df.select(
                        self.user_col, F.col(self.item_col).alias("train_item_tmp")
                    ),
                    on=[self.user_col],
                )
                .select(
                    self.user_col,
                    self.item_col,
                    F.least(F.col("reco_item_tmp"), F.col("train_item_tmp")).alias(
                        "i1"
                    ),
                    F.greatest(F.col("reco_item_tmp"), F.col("train_item_tmp")).alias(
                        "i2"
                    ),
                )
                .join(self.df_cosine_similariy, on=["i1", "i2"], how="left")
                .fillna(0)
                .groupBy(self.user_col, self.item_col)
```

Reply:
adopted the change.